src/core/xref.js

```diff
@@ -164,150 +110,117 @@ function normalizeConfig(xref) {
-    if (isIDL) {
-      if (elem.dataset.xrefType) {
-        types.push(...elem.dataset.xrefType.split("|"));
-      } else {
-        types.push("_IDL_");
-      }
+  const types = [];
+  if (isIDL) {
+    if (elem.dataset.xrefType) {
+      types.push(...elem.dataset.xrefType.split("|"));
     } else {
-      types.push("_CONCEPT_");
+      types.push("_IDL_");
     }
+  } else {
+    types.push("_CONCEPT_");
+  }
 
-    let { linkFor: forContext = null } = elem.dataset;
-    if (!forContext && isIDL) {
-      const dataLinkForElem = elem.closest("[data-link-for]");
-      if (dataLinkForElem) {
-        forContext = dataLinkForElem.dataset.linkFor;
-      }
+  let { linkFor: forContext } = elem.dataset;
+  if (!forContext && isIDL) {
+    const dataLinkForElem = elem.closest("[data-link-for]");
+    if (dataLinkForElem) {
+      forContext = dataLinkForElem.dataset.linkFor;
     }
-
-    const xrefsForTerm = map.has(term) ? map.get(term) : [];
-    xrefsForTerm.push({ elem, specs: uniqueSpecs, for: forContext, types });
-    map.set(term, xrefsForTerm);
   }
 
-  return map;
+  return {
+    term,
+    types,
+    ...(specs.length && { specs: [...new Set(specs)].sort() }),
+    ...(forContext && { for: forContext }),
+  };
 }
 
 /**
- * collects xref keys in a form more usable for querying
- * @param {ReturnType<typeof createXrefMap>} xrefs
- * @typedef {{ term: string, specs?: string[], types?: string[], for?: string }} PayloadKey
- * @returns {PayloadKey[]}
+ * @param {RequestEntry[]} queryKeys
+ * @param {string} apiUrl
+ * @returns {Promise<Map<string, SearchResultEntry[]>>}
  */
-function collectPayloadKeys(xrefs) {
-  const keys = new ObjectSet();
-  for (const [term, entries] of xrefs) {
-    for (const { specs, types, for: forContext } of entries) {
-      const key = { term };
-      if (specs.length) key.specs = specs;
-      if (types.length) key.types = types;
-      if (forContext) key.for = forContext;
-      keys.add(key);
-    }
+async function getData(queryKeys, apiUrl) {
+  const idb = await openDB("xref", 1, {
+    upgrade(db) {
+      db.createObjectStore("xrefs");
+    },
+  });
+  const cache = new IDBKeyVal(idb, "xrefs");
+  const resultsFromCache = await resolveFromCache(queryKeys, cache);
+
+  const termsToLook = queryKeys.filter(key => !resultsFromCache.get(key.hash));
+  const fetchedResults = await fetchFromNetwork(termsToLook, apiUrl);
+  if (fetchedResults.size) {
+    // add data to cache
+    await cache.addMany(fetchedResults);
+    await cache.set("__CACHE_TIME__", Date.now());
   }
-  return [...keys];
-}
 
-// adds data to cache
-async function cacheResults(data, cache) {
-  const promisesToSet = Object.entries(data).map(([term, results]) =>
-    cache.set(term, results)
-  );
-  await cache.set("__CACHE_TIME__", Date.now());
-  await Promise.all(promisesToSet);
+  // merge above two Maps. faster than following equivalent:
+  // return new Map([...resultsFromCache, ...fetchedResults]);
```

Comment:
I think you should do the: 

```JS
 return new Map([...resultsFromCache, ...fetchedResults]);
```

Let engines catch up on perf. Don't work around it. It's also going to be a tiny per hit.